src/commands/build-util.ts

```diff
@@ -1,19 +1,126 @@
+import path from 'path';
+import {SnowpackBuildMap, SnowpackConfig, SnowpackPlugin, SnowpackSourceFile} from '../config';
+import {getExt, replaceExt} from '../util';
+
+export interface BuildFileOptions {
+  buildPipeline: SnowpackPlugin[];
+  messageBus: EventEmitter;
+  isDev: boolean;
+}
 
 const IS_PREACT = /from\s+['"]preact['"]/;
 export function checkIsPreact(filePath: string, contents: string) {
   return filePath.endsWith('.jsx') && IS_PREACT.test(contents);
 }
 
+export function getInputsFromOutput(fileLoc: string, plugins: SnowpackPlugin[]) {
+  const ext = path.extname(fileLoc);
+  const potentialInputs = new Set([fileLoc]);
+  for (const plugin of plugins) {
+    if (plugin.output.includes(ext)) {
+      plugin.input.forEach((inp) => potentialInputs.add(fileLoc.replace(ext, inp)));
+    }
+  }
+  return Array.from(potentialInputs);
+}
+
+/** Core Snowpack file pipeline builder */
+export async function buildFile(
+  srcPath: string,
+  srcContents: string,
+  {buildPipeline, messageBus, isDev}: BuildFileOptions,
+): Promise<SnowpackBuildMap> {
+  const srcExt = getExt(srcPath);
+  const output: SnowpackBuildMap = {}; // important: clear output for each src file to keep memory low
+  output[srcPath] = {...srcExt, contents: srcContents, locOnDisk: srcPath}; // this is the object we’ll mutate through transformations
+
+  // Run the file through the Snowpack build pipeline:
+  for (const step of buildPipeline) {
+    // For now, we only run through build plugins that match at least one file extension. 
+    // All other plugins are ignored.
+    if (!step.build || step.input.length === 0) {
+      continue;
+    }
+    const staleOutputs = new Set<string>();
+    for (const destPath of Object.keys(output)) {
+      // If the current output file extension doesn't match a plugin input, skip it.
+      if (!step.input.includes(getExt(destPath).baseExt)) {
+        continue;
+      }
+      const destBuildFile = output[destPath];
+      const result = await step.build({
+        contents: destBuildFile.contents,
+        filePath: destPath,
+        isDev,
+        log: (msg, data) => {
+          messageBus.emit(msg, {
+            ...data,
+            id: step.name,
+            msg: data.msg && `[${destPath}] ${data.msg}`,
+          });
+        },
+        // Deprecated
+        urlPath: `./${path.basename(destPath)}`,
+      });
+      if (!result) {
+        continue;
+      }
+      if (typeof result === 'string') {
+        // Path A: single-output (assume extension is same)
+        destBuildFile.contents = result;
+      } else if (typeof result === 'object' && !result.result) {
+        // Path B: multi-file output ({ js: [string], css: [string], … })
+        Object.entries(result as {[ext: string]: string}).forEach(([ext, contents]) => {
+          if (!contents) {
+            return;
+          }
+          const newDest = replaceExt(destPath, ext);
+          output[newDest] = {
+            baseExt: ext,
```

Comment:
Should `baseExt` and `expandedExt` be generated and spread by the `getExt(newDest)` since they do a bit more logic than just put the ext as those properties?

Reply:
See comment above. I think this data representation in this function won't last long after this PR, I'd prefer to spend the time elsewhere